packages/lesswrong/server/partiallyReadSequences.js

```diff
@@ -0,0 +1,151 @@
+      numTotal: postIDs.length
+    };
+    
+    // Generate a new partiallyReadSequences list by filtering out any previous
+    // entry for this sequence or for the collection that contains it, and
+    // adding a new entry for this sequence to the end.
+    const newPartiallyReadSequences = [...partiallyReadMinusThis, sequenceReadStatus];
+    setUserPartiallyReadSequences(userId, newPartiallyReadSequences);
+    return;
+  }
+  
+  // User is finished with this sequence. If the sequence had a canonical
+  // collection (ie, R:A-Z, Codex or HPMoR), find the first unread post in
+  // the whole collection. If they've read everything in the collection, or
+  // it isn't part of a collection, they're done.
+  if (collection) {
+    const collectionPostIDs = await Collections.getAllPostIDs(collection._id);
+    const collectionPostReadStatuses = await postsToReadStatuses(user, collectionPostIDs);
+    const collectionAnyUnread = _.some(collectionPostIDs, postID => !collectionPostReadStatuses[postID]);
+    
+    if (collectionAnyUnread) {
+      const nextPostIndex = findIndex(collectionPostIDs, postID=>!collectionPostReadStatuses[postID]);
+      const nextPostId = collectionPostIDs[nextPostIndex];
+      
+      const collectionReadStatus = {
+        collectionId: collection._id,
+        lastReadPostId: postId,
+        nextPostId: nextPostId,
+        numRead: _.filter(collectionPostIDs, id=>collectionPostReadStatuses[id]).length,
+        numTotal: collectionPostIDs.length
+      };
+      
+      // Generate a new partiallyReadSequences, filtering out the sequence that
+      // was just finished and any other entry for this same collection.
+      //
+      // Minor oddity: It's possible to end up with the partially-read
+      // list containing both a sequence and the collection that contains it,
+      // if you are part-way through sequence A, and finish sequence B, A and
+      // B in the same collection.
+      const newPartiallyReadSequences = [...partiallyReadMinusThis, collectionReadStatus];
+      setUserPartiallyReadSequences(userId, newPartiallyReadSequences);
+      return;
+    }
+  }
+  
+  // Done reading! If the user previously had a partiallyReadSequences entry
+  // for this sequence, remove it and update the user object.
+  if (_.some(user.partiallyReadSequences, s=>s.sequenceId === sequenceId)) {
+    setUserPartiallyReadSequences(userId, partiallyReadMinusThis);
+  }
+}
+
+export const setUserPartiallyReadSequences = async (userId, newPartiallyReadSequences) => {
+  editMutation({
+    collection: Users,
+    documentId: userId,
+    set: {
+      partiallyReadSequences: newPartiallyReadSequences
+    },
+    unset: {},
+    validate: false,
+  });
+}
+
+const EventUpdatePartialReadStatusCallback = async (event) => {
+  if (event.name === 'post-view' && event.properties.sequenceId) {
+    // Deliberately lacks an await - this runs concurrently in the background
```

Comment:
But then why is this function async? To make the type signature better I would make this function sync, since it will never return a promise.